### PR TITLE
[AMDGPU] Still set up the two SGPRs for queue ptr even it is COV5

### PR DESCRIPTION
**Pull request by shiltian**

### CODE REVIEW

**Comment by shiltian** [comment]
* **#112403** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/112403?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @shiltian and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: llvm/test/CodeGen/AMDGPU/andorbitset.ll [hunk 1]
```diff
@@ -4,12 +4,12 @@
...
 ; SI:       ; %bb.0:
-; SI-NEXT:    s_load_dword s4, s[2:3], 0xb
-; SI-NEXT:    s_load_dwordx2 s[0:1], s[2:3], 0x9
+; SI-NEXT:    s_load_dword s6, s[4:5], 0xb
+; SI-NEXT:    s_load_dwordx2 s[0:1], s[4:5], 0x9
 ; SI-NEXT:    s_mov_b32 s3, 0xf000
 ; SI-NEXT:    s_mov_b32 s2, -1
 ; SI-NEXT:    s_waitcnt lgkmcnt(0)
-; SI-NEXT:    s_bitset0_b32 s4, 31
+; SI-NEXT:    s_and_b32 s4, s6, 0x7fffffff
```
Comment (shiltian) [inline llvm/test/CodeGen/AMDGPU/andorbitset.ll:12]: @arsenm This looks like a big regression. Not sure why that happens.

Comment (arsenm) [reply llvm/test/CodeGen/AMDGPU/andorbitset.ll:12]: It's only a code size regression. The register allocation changed so now the result isn't the same as the input register. We could add a regalloc hint somewhere to prefer keeping source = dest

**Comment by shiltian** [comment]
The only test failure is `CodeGen/AMDGPU/call-args-inreg.ll`. It crashes when lowering function `test_call_external_void_func_a15i32_inreg`.

https://github.com/llvm/llvm-project/blob/8d13e7b8c382499c1cf0c2a3184b483e760f266b/llvm/lib/Target/AMDGPU/SIFrameLowering.cpp#L905

File: llvm/test/CodeGen/AMDGPU/andorbitset.ll (hunk 1 above)
Comment (shiltian) [reply llvm/test/CodeGen/AMDGPU/andorbitset.ll:12]: Okay, then I guess we can just leave as it is for now.

**Comment by arsenm** [comment]
> The only test failure is `CodeGen/AMDGPU/call-args-inreg.ll`. It crashes when lowering function `test_call_external_void_func_a15i32_inreg`.
> 
> https://github.com/llvm/llvm-project/blob/8d13e7b8c382499c1cf0c2a3184b483e760f266b/llvm/lib/Target/AMDGPU/SIFrameLowering.cpp#L905

Something is wrong with how inreg is interacting with special arguments at callsites. The special arguments shouldn't be changing the number of registers available for user arguments.

**Comment by shiltian** [comment]
yeah we have this issue the first day we support inreg SGPR argument (https://github.com/llvm/llvm-project/pull/67182) and at that moment COV5 was already the default so the issue was not exposed.

**Comment by arsenm** [comment]
> The fix might be to spill one SGPR or reserve one ahead of time when allocating SGPRs for arguments.

We probably should just reserved a register always for these spill situations. The SGPR argument part is incidental

**Comment by shiltian** [comment]
Created https://github.com/llvm/llvm-project/issues/113782 to track the issue.

**Review by arsenm — APPROVED** [review]

**Comment by arsenm** [comment]
> https://lab.llvm.org/buildbot/#/builders/123/builds/9031

That's one of the flaky tests that randomly fails

**Comment by jplehr** [comment]
> > https://lab.llvm.org/buildbot/#/builders/123/builds/9031
> 
> That's one of the flaky tests that randomly fails

I thought the one that was failing randomly for some time now has been disabled and the HIP bot is still red.

All OpenMP bots are also red.
https://lab.llvm.org/buildbot/#/builders/30
https://lab.llvm.org/buildbot/#/builders/73
https://lab.llvm.org/staging/#/builders/105
https://lab.llvm.org/staging/#/builders/97
https://lab.llvm.org/staging/#/builders/130

**Comment by shiltian** [comment]
This change is likely to expose an issue in the register allocation, as we now lose 2 SGPRs by default.

**Comment by kosarev** [comment]
@shiltian This seems to unintentionally add the binary file llvm/test/CodeGen/AMDGPU/amdhsa-kernarg-preload-num-sgprs.o to tests.

**Comment by shiltian** [comment]
> @shiltian This seems to unintentionally add the binary file llvm/test/CodeGen/AMDGPU/amdhsa-kernarg-preload-num-sgprs.o to tests.

Yes, indeed. Thanks! I fixed it downstream and I thought it was added there.

**Pull request merged**